### PR TITLE
🐛 ec2: instances: fix assigning public IP

### LINKED ISSUE #4888: No validation done when only `PublicIP == true` but public subnet has `MapPublicIPOnLauch == false`
/kind bug

**What steps did you take and what happened:**
1. Create a shared VPC with private and public subnets, where all public subnets have `MapPublicIPOnLaunch == false`.
2. Create an AWSMachine with `PublicIP = true`  and `Subnet = nil`

**What did you expect to happen:**
CAPA throws an error letting the user know that no public IP will be assigned to the instance.

Instead what happens is the machine only has a private IP assigned and cannot talk to the internet because the internet gateway doesn't provide NAT.

**Anything else you would like to add:**
https://cluster-api-aws.sigs.k8s.io/topics/bring-your-own-aws-infrastructure does not say that having `MapPublicIPOnLaunch == true` is a requirement for public subnets, even though it expects that. Note that the condition is checked if a subnet ID is set in the AWSMachine spec: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/ec2/instances.go#L337-L340

**Environment:**

- Cluster-api-provider-aws version: main @ d62768f507f85b25db1e5a74e0e820ed1f8e4414
- Kubernetes version: (use `kubectl version`): 1.29
- OS (e.g. from `/etc/os-release`): RHCOS

### PR DESCRIPTION
**Pull request by r4f4**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

In the scenario where the user brings their own VPC, if no subnet ID is set in the machine spec and PublicIP is true, CAPA will choose one from the available public subnets. However, if the subnet doesn't have MapPublicIPOnLaunch == true, the instance will not be assigned a public IP. As a result, the instance will have no internet access, contrary to the user's expectation.

This change guarantees an instance will be assigned a public IP even if the subnet doesn't do it on instance launch. Instead, we set the option in the instance's network interface.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4888

**Special notes for your reviewer**:

I still need to fix and add unit tests. I want to get feedback on the approach first.

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [X] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Fix machines getting a public IP even when user-supplied subnets don't have MapPublicIpOnLaunch.
```

### CODE REVIEW

**Comment by r4f4** [comment]
/hold
Until I fix the unit tests.

File: pkg/cloud/services/ec2/instances.go [hunk 1]
```diff
@@ -334,7 +359,7 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
 					*subnet.SubnetId, *subnet.AvailabilityZone, *failureDomain)
 				continue
 			}
-			if scope.AWSMachine.Spec.PublicIP != nil && *scope.AWSMachine.Spec.PublicIP && !*subnet.MapPublicIpOnLaunch {
+			if wantsPublicIP && !isSubnetPublic(routeTables, *subnet.SubnetId) {
```
Comment (mtulio) [inline pkg/cloud/services/ec2/instances.go:362]: @r4f4 can we use the `IsPublic` flag for the public subnet check?

```suggestion
			if wantsPublicIP && subnet.IsPublic {
```

Comment (r4f4) [reply pkg/cloud/services/ec2/instances.go:362]: @mtulio not blindly. At this point all we have is an `ec2.Subnet` which doesn't have a simple `IsPublic` field. We might be able to do that if we separate the case subnetID-supplied from the subnetFilter-supplied.

Comment (mtulio) [reply pkg/cloud/services/ec2/instances.go:362]: @r4f4 apologies, I should have linked the field :)

The [SubnetSpec.IsPublic](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/f67715ea3c17728fc531ec488fb0e186f31b2457/api/v1beta2/network_types.go#L413-L416) is filled when the subnet is created. Can you access in this stage the flag from the subnet ID from `subnet.SubnetId`? Something like:
```sh
s.scope.Subnets().FindByID(*subnet.SubnetId).IsPublic
```

Comment (r4f4) [reply pkg/cloud/services/ec2/instances.go:362]: Aha! That makes sense. Nice suggestion, will incorporate!

Comment (r4f4) [reply pkg/cloud/services/ec2/instances.go]: Done

**Comment by damdo** [comment]
/cc. @vincepri

**Comment by damdo** [comment]
@r4f4 something needs fixing in API generation and units

**Comment by r4f4** [comment]
> @r4f4 something needs fixing in API generation and units. Could you PTAL? Thanks!

Sure, I also need to add unit tests.

**Comment by r4f4** [comment]
@damdo I don't know how to fix the fuzzy test. Would you mind giving some pointers?

**Comment by damdo** [comment]
@r4f4 The Fuzzy testing in this instance refers to the Conversion across API types conversion.
More specifically `--- FAIL: TestFuzzyConversion/for_AWSCluster` refers to the AWSCluster conversion.

Try and add 
```
dst.Status.Bastion.PublicIPOnLaunch = restored.Status.Bastion.PublicIPOnLaunch
```
after this: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/daaa1e12a8065329a5f8f2d10c207b7b739d9538/api/v1beta1/awscluster_conversion.go#L59

**Comment by r4f4** [comment]
@damdo thanks! That did the trick.

**Review by damdo — COMMENTED** [review]
Code and reasoning look reasonable.
E2Es passed.

/lgtm

/cc @nrb 
/assign @richardcase @dlipovetsky @vincepri

**Comment by r4f4** [comment]
I'll rebase and squash the commits.

**Review by damdo — COMMENTED** [review]
/lgtm

File: pkg/cloud/services/ec2/instances.go [hunk 2]
```diff
@@ -538,13 +553,25 @@ func (s *Service) runInstance(role string, i *infrav1.Instance) (*infrav1.Instan
 				DeviceIndex:        aws.Int64(int64(index)),
 			})
 		}
+		netInterfaces[0].AssociatePublicIpAddress = i.PublicIPOnLaunch
 
 		input.NetworkInterfaces = netInterfaces
 	} else {
-		input.SubnetId = aws.String(i.SubnetID)
+		if i.PublicIPOnLaunch != nil && *i.PublicIPOnLaunch {
```
Comment (vincepri) [inline pkg/cloud/services/ec2/instances.go:560]: Use ptr.Deref?

File: api/v1beta2/types.go [hunk 3]
```diff
@@ -237,6 +237,10 @@ type Instance struct {
 	// PrivateDNSName is the options for the instance hostname.
 	// +optional
 	PrivateDNSName *PrivateDNSName `json:"privateDnsName,omitempty"`
+
+	// PublicIPOnLaunch is the option to associate a public IP on instance launch
+	// +optional
+	PublicIPOnLaunch *bool `json:"publicIPOnLaunch,omitempty"`
```
Comment (vincepri) [inline api/v1beta2/types.go:243]: Would this field now allow to set a public ip on a Machine in a private subnet?

Comment (r4f4) [reply api/v1beta2/types.go:243]: I think so, although I haven't tested it.

File: pkg/cloud/services/ec2/instances.go (hunk 2 above)
Comment (r4f4) [reply pkg/cloud/services/ec2/instances.go]: Done.

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by vincepri** [comment]
@r4f4 squash commits?

**Review by damdo — COMMENTED** [review]
/lgtm

**Comment by damdo** [comment]
@vincepri @richardcase @dlipovetsky @Ankitasw this one needs an `/approve`

**Comment by richardcase** [comment]
Until the e2e pass:

/hold

**Comment by richardcase** [comment]
But from my side:

/approve

When the e2e passes feel free to unhold.

**Comment by damdo** [comment]
It timed out
/test pull-cluster-api-provider-aws-e2e

**Comment by r4f4** [comment]
Are we good to remove the hold?

**Pull request merged**